crates/dbsp/src/operator/dynamic/balance/accumulate_trace_balanced.rs

```diff
@@ -0,0 +1,1306 @@
+                    if cursor.val_valid() {
+                        break;
+                    }
+                    cursor.step_key();
+                }
+            }
+            PartitioningPolicy::Balance => {
+                let mut has_values = vec![false; shards];
+                let mut shard = 0;
+
+                while cursor.key_valid() {
+                    while cursor.val_valid() && builders[shard].num_tuples() < chunk_size {
+                        let mut weight = 0;
+                        cursor.map_times(&mut |_t, w| {
+                            weight += **w;
+                        });
+                        if weight != 0 {
+                            shard =
+                                Self::hash_key_val(cursor.key(), cursor.val()) as usize % shards;
+                            let b = &mut builders[shard];
+
+                            has_values[shard] = true;
+                            b.push_val_diff(cursor.val(), &weight);
+                        }
+                        cursor.step_val();
+                    }
+
+                    for (shard, builder) in builders.iter_mut().enumerate() {
+                        if has_values[shard] {
+                            builder.push_key(cursor.key());
+                        }
+                    }
+                    if cursor.val_valid() {
+                        break;
+                    }
+                    cursor.step_key();
+                    has_values.fill(false);
+                }
+            }
+            PartitioningPolicy::Broadcast => {
+                while cursor.key_valid() {
+                    let mut has_values = false;
+
+                    while cursor.val_valid() && builders[0].num_tuples() < chunk_size {
+                        let mut weight = 0;
+                        cursor.map_times(&mut |_t, w| {
+                            weight += **w;
+                        });
+                        for b in builders.iter_mut() {
+                            if weight != 0 {
+                                has_values = true;
+                                b.push_val_diff(cursor.val(), &weight);
+                            }
+                        }
+                        cursor.step_val();
+                    }
+                    if has_values {
+                        for b in builders.iter_mut() {
+                            b.push_key(cursor.key());
+                        }
+                    }
+                    if cursor.val_valid() {
+                        break;
+                    }
+                    cursor.step_key();
+                }
+            }
+        }
+    }
+
+    /// Partition cursor based on the current policy.
+    ///
+    /// Assumes that the cursor was partitioned using Policy::Broadcast, i.e., all partitions contain identical data.
+    /// The implementation takes advantage of this by having each worker only send data that belongs to the same worker
```

Comment:
If they have already the data, why is sending even needed? Don't they just have to delete some of the data they already have?

Reply:
Are you doing this because it's cheaper to retract everything and send what is needed because the data structures are immutable? Still, you don't need to "send" anything, the data is already there, every worker should be able to compute what to keep.